File: backend/app/games/minion_clash/js/battle/SpawnSystem.js
import { Unit } from '../entities/Unit.js';
import { GameConfig } from '../config/GameConfig.js';

/**
 * SpawnSystem: enqueues unit spawns from played cards and resolves them on
 * next tick (so spawn happens after the current update pass).
 */
export class SpawnSystem {
    constructor(entityManager, dataRegistry, modifiers, assets) {
        this._em = entityManager;
        this._data = dataRegistry;
        this._modifiers = modifiers ?? { enemyUnitHpMult: 1, playerUnitHpMult: 1 };
        this._assets = assets ?? null;
        this._queue = [];
    }

    /**
     * Enqueue summon from a card definition.
     * @param {object} cardDef
     * @param {'player'|'enemy'} team
     * @param {number} x @param {number} y
     */
    enqueueFromCard(cardDef, team, x, y) {
        if (cardDef.kind !== 'summon') {
            throw new Error(`SpawnSystem: card "${cardDef.id}" is not a summon`);
        }
        this._queue.push({ team, unitId: cardDef.unitId, count: cardDef.count, spread: cardDef.spread, x, y });
    }

    /** Direct spawn used by onDeath triggers (e.g. Necromancer). */
    spawnUnitsAt(team, unitId, count, x, y, spread = 18) {
        this._queue.push({ team, unitId, count, spread, x, y });
    }

    flush() {
        if (this._queue.length === 0) return;
        for (const job of this._queue) this._spawnJob(job);
        this._queue.length = 0;
    }

    _spawnJob(job) {
        const def = this._data.getUnit(job.unitId);
        const scaled = this._scaleDefinition(def, job.team);
        const positions = this._cluster(job.x, job.y, job.count, job.spread);
        const spriteRef = scaled.sprite ?? { prefix: `unit_${job.unitId}`, scale: 1 };
        for (const p of positions) {
            const unit = new Unit({ team: job.team, x: p.x, y: p.y, def: scaled });
            if (this._assets) {
                unit.setSprite(this._assets.createAnimator(spriteRef));
            }
            this._em.add(unit);
        }
    }

    _scaleDefinition(def, team) {
        const mult = team === 'enemy' ? this._modifiers.enemyUnitHpMult : this._modifiers.playerUnitHpMult;
        if (!mult || mult === 1) return def;
        return Object.freeze({ ...def, hp: Math.round(def.hp * mult) });
    }

    _cluster(x, y, count, spread) {
        if (count <= 1) return [this._clamp(x, y)];
        const out = [];
        const step = (Math.PI * 2) / count;
        for (let i = 0; i < count; i++) {
            const a = step * i;
            out.push(this._clamp(x + Math.cos(a) * spread, y + Math.sin(a) * spread));
        }
        return out;
    }

    _clamp(x, y) {
        const a = GameConfig.ARENA;
        return {
            x: Math.max(20, Math.min(GameConfig.VIEW_WIDTH - 20, x)),
            y: Math.max(a.TOP + 10, Math.min(a.BOTTOM - 10, y))
        };
    }
}

File: backend/app/games/minion_clash_be/simulation/spawn_system.py
"""
SpawnSystem — enqueues unit spawns from played cards and resolves them
on next tick (so spawn happens after the current update pass).

Mirrors `SpawnSystem.js` for `summon` cards. Cluster pattern is
deterministic (angle = step * i).
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

from . import arena
from .data_loader import DataRegistry
from .entities import Unit
from .entity_manager import EntityManager


@dataclass
class _SpawnJob:
    team: str
    unit_id: str
    count: int
    spread: float
    x: float
    y: float


class SpawnSystem:
    def __init__(self, em: EntityManager, data: DataRegistry):
        self._em = em
        self._data = data
        self._queue: list[_SpawnJob] = []
        self._spawned_this_flush: list[Unit] = []

    def enqueue_from_card(self, card: dict[str, Any], team: str, x: float, y: float) -> None:
        if card.get("kind") != "summon":
            raise ValueError(f"SpawnSystem: card {card.get('id')!r} is not a summon")
        self._queue.append(_SpawnJob(
            team=team,
            unit_id=card["unitId"],
            count=int(card.get("count", 1)),
            spread=float(card.get("spread", 18)),
            x=x,
            y=y,
        ))

    def enqueue_unit(
        self, *, team: str, unit_id: str, x: float, y: float,
        count: int = 1, spread: float = 18.0,
    ) -> None:
        """Direct unit spawn (used by onDeath triggers)."""
        self._queue.append(_SpawnJob(
            team=team, unit_id=unit_id,
            count=max(1, int(count)), spread=float(spread),
            x=float(x), y=float(y),
        ))

    def flush(self) -> list[Unit]:
        """Resolve queue. Returns the list of units spawned this flush."""
        self._spawned_this_flush = []
        if not self._queue:
            return self._spawned_this_flush
        for job in self._queue:
            self._resolve(job)
        self._queue.clear()
        return self._spawned_this_flush

    def _resolve(self, job: _SpawnJob) -> None:
        u_def = self._data.get_unit(job.unit_id)
        for px, py in self._cluster(job.x, job.y, job.count, job.spread):
            unit = Unit(team=job.team, x=px, y=py, def_dict=u_def)
            self._em.add(unit)
            self._spawned_this_flush.append(unit)

    @staticmethod
    def _cluster(x: float, y: float, count: int, spread: float) -> list[tuple[float, float]]:
        if count <= 1:
            return [SpawnSystem._clamp(x, y)]
        out: list[tuple[float, float]] = []
        step = (math.pi * 2) / count
        for i in range(count):
            a = step * i
            out.append(SpawnSystem._clamp(
                x + math.cos(a) * spread,
                y + math.sin(a) * spread,
            ))
        return out

    @staticmethod
    def _clamp(x: float, y: float) -> tuple[float, float]:
        return (
            max(20.0, min(arena.VIEW_WIDTH - 20.0, x)),
            max(arena.ARENA_TOP + 10.0, min(arena.ARENA_BOTTOM - 10.0, y)),
        )
